### PR TITLE
[issue1138] Raise exception in exit_with() to ensure that destructors are called upon program exit.

### PR DESCRIPTION
**Pull request by jendrikseipp**

### CODE REVIEW

File: src/search/utils/system.cc [hunk 1]
```diff
@@ -46,7 +46,7 @@ bool is_exit_code_error_reentrant(ExitCode exitcode) {
 
 void exit_with(ExitCode exitcode) {
     report_exit_code_reentrant(exitcode);
```
Comment (maltehelmert) [inline src/search/utils/system.cc:48]: It's a drive-by change, but while we're looking at this: can we change this call to a new function `report_exit_code` that plays well with iostreams? At the moment this is likely to break output if we don't flush manually before calling `exit_with`.

Comment (jendrikseipp) [reply src/search/utils/system.cc]: Done.

File: src/search/utils/system.cc [hunk 2]
```diff
@@ -46,7 +46,7 @@ bool is_exit_code_error_reentrant(ExitCode exitcode) {
...
-    exit(static_cast<int>(exitcode));
+    throw ExitException(exitcode);
```
[the remark this answers is not in the document]
Comment (maltehelmert) [reply src/search/utils/system.cc:62]: Agreed with Jendrik, in my view the try/catch block in the main function is part of the overall program design. And if we assume it as given, exit_with *does* make these guarantees. I don't think it's unusual that a piece of code needs other code to cooperate in order to satisfy its contract, although admittedly it's slightly unusual that the dependency takes this form. We might add comments for one or both of the try/catch block and the exit_with function, but for me it's not necessary.

To be clear, for me the whole existence of this exception class is an implementation detail meant to be hidden from callers of exit_with. It would be wrong (in the sense of breaking an intended encapsulation) to create or catch ExitException anywhere else.

[the remark this answers is not in the document]
Comment (FlorianPommerening) [reply src/search/utils/system.cc:62]: I'm fine with the change, I just thought about renaming the function. But I don't have a better suggestion, so we can leave it as is. I guess the idea then is "calling this function will eventually lead to an exit()" rather than "this function will exit()".
